### PR TITLE
Implement #4479

### PR DESCRIPTION
**Pull request by X123M3-256**
This PR adds a routine which will check whether a vehicle has sprites for the track pieces that it claims to support in the DAT. If it doesn't, it will act as if the piece was disabled in the DAT, i.e you can't build the piece in the ride construction menu.

The primary reason for having this is that many vehicles in the game enable all track pieces, and rely on the limitations of the track type itself. As these track types are extended with more pieces, it becomes possible to build a ride that will glitch without using any cheats, which is IMO undesirable.

Unfortunately, if the ride enables diagonal slopes but does not have the sprites, this won't detect it. That's because you can't suppress diagonal slopes seperately - you'd have to disable either diagonals in general, or slopes in general. Obviously, I do not want this to disable pieces that do have correct sprites, so in this case it lets through the missing pieces. Luckily, the issues caused by missing diagonal sprites are relatively subtle.

I feel that this shouldn't be merged until there is a cheat to disable track piece restrictions entirely, as otherwise it would be excessively limiting when you are deliberately using a train with a different track type (many people will put up with the glitches for the extra pieces). I will work on that next.

### CODE REVIEW

File: src/ride/ride.c [hunk 1]
```diff
@@ -7457,6 +7457,21 @@ uint8 ride_entry_get_vehicle_at_position(int rideEntryIndex,int numCarsPerTrain,
...
+
+//Finds track pieces that a given ride entry doesn't have the sprites for
+uint64 ride_entry_get_supported_track_pieces(rct_ride_entry* rideEntry)
+{
+uint64 supportedPieces = 0xFFFFFFFFFFFFFFFFULL;
+uint16 trackPieceRequiredSprites[55] = {0x0001u,0x0001u,0x0001u,0x0000u,0x0006u,0x0002u,0x0020u,0x000E,0x0003u,0x0006u,0x0007u,0x0002u,0x0004u,0x0001u,0x0001u,0x0001u,0x0001u,0x0061u,0x000E,0x1081u,0x0001u,0x0020u,0x0020u,0x0001u,0x0001u,0x0000u,0x0001u,0x0001u,0x000C,0x0061u,0x0002u,0x000E,0x0480u,0x0001u,0x0061u,0x0001u,0x0001u,0x000Fu,0x0001u,0x0200u,0x0007u,0x0008u,0x0000u,0x0000u,0x4000u,0x0008u,0x0001u,0x0001u,0x0061u,0x0061u,0x0008u,0x0008u,0x0001u,0x000Eu,0x000Eu};
+//Only check default vehicle; it's assumed the others will have correct sprites if this one does (I've yet to find an exception, at least)
+		for(int j = 0; j < 55; j++)
+		{
+			if((rideEntry->vehicles[rideEntry->default_vehicle].sprite_flags & trackPieceRequiredSprites[j]) != trackPieceRequiredSprites[j])supportedPieces &= ~(1ULL << j);
```
Comment (Gymnasiast) [inline src/ride/ride.c]: The code does not compile.

At the very least there is a left brace missing here.

File: src/ride/ride.c [hunk 2]
```diff
@@ -7467,7 +7467,6 @@ uint16 trackPieceRequiredSprites[55] = {0x0001u,0x0001u,0x0001u,0x0000u,0x0006u,
 		for(int j = 0; j < 55; j++)
 		{
 			if((rideEntry->vehicles[rideEntry->default_vehicle].sprite_flags & trackPieceRequiredSprites[j]) != trackPieceRequiredSprites[j])supportedPieces &= ~(1ULL << j);
```
Comment (AaronVanGeffen) [inline src/ride/ride.c]: Could you introduce a few newline here for legibility?

**Comment by Gymnasiast** [comment]
Please mind the code style next time. It's important to keep the code readable. I fixed it this time.

Originally, I was going to merge this after 0.0.5, but since so much stuff got in I'll just test it thoroughly and merge it if it doesn't cause problems.

File: src/ride/ride.c [hunk 3]
```diff
@@ -7457,6 +7457,22 @@ uint8 ride_entry_get_vehicle_at_position(int rideEntryIndex,int numCarsPerTrain,
 	}
 }
 
+
+//Finds track pieces that a given ride entry doesn't have the sprites for
+uint64 ride_entry_get_supported_track_pieces(rct_ride_entry* rideEntry)
+{
+	uint64 supportedPieces = 0xFFFFFFFFFFFFFFFFULL;
+	uint16 trackPieceRequiredSprites[55] = {0x0001u,0x0001u,0x0001u,0x0000u,0x0006u,0x0002u,0x0020u,0x000E,0x0003u,0x0006u,0x0007u,0x0002u,0x0004u,0x0001u,0x0001u,0x0001u,0x0001u,0x0061u,0x000E,0x1081u,0x0001u,0x0020u,0x0020u,0x0001u,0x0001u,0x0000u,0x0001u,0x0001u,0x000C,0x0061u,0x0002u,0x000E,0x0480u,0x0001u,0x0061u,0x0001u,0x0001u,0x000Fu,0x0001u,0x0200u,0x0007u,0x0008u,0x0000u,0x0000u,0x4000u,0x0008u,0x0001u,0x0001u,0x0061u,0x0061u,0x0008u,0x0008u,0x0001u,0x000Eu,0x000Eu};
```
Comment (janisozaur) [inline src/ride/ride.c]: please break it to multiple lines

**Comment by Gymnasiast** [comment]
I have merged the detection part, as we're sure about the conditionals yet.

**Pull request closed without merging**